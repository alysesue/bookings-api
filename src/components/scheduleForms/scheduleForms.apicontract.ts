export class ScheduleFormRequest {
	/**
	 * @maxLength 100
	 */
	public name: string;
	public serviceProviderId?: number;

	/**
	 * @isInt
	 * @minimum 1
	 * @maximum 480
	 */
	public slotsDurationInMin: number;
	public weekdaySchedules: WeekDayScheduleContract[];
}

export class WeekDayBreakContract {
	/**
	 * @maxLength 5
	 */
	public startTime: string;
	/**
	 * @maxLength 5
	 */
	public endTime: string;
}

export class WeekDayScheduleContract {
	/**
	 * @isInt
	 */
	public weekday: number;
	public hasScheduleForm: boolean;
	/**
	 * @maxLength 5
	 */
	public openTime?: string;
	/**
	 * @maxLength 5
	 */
	public closeTime?: string;
	public breaks: WeekDayBreakContract[];
}

export class ScheduleFormResponse {
	/**
	 * @isInt
	 */
	public id: number;
	public name: string;
	/**
	 * @isInt
	 */
	public slotsDurationInMin: number;
	public weekdaySchedules: WeekDayScheduleContract[];
}
