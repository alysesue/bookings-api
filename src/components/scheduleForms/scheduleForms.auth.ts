// tslint:disable: tsr-detect-possible-timing-attacks
import {
	CitizenAuthGroup,
	OrganisationAdminAuthGroup,
	ServiceAdminAuthGroup,
	ServiceProviderAuthGroup,
} from '../../infrastructure/auth/authGroup';
import { ServiceProvider } from '../../models';
import {
	PermissionAwareAuthGroupVisitor,
	QueryAuthGroupVisitor,
} from '../../infrastructure/auth/queryAuthGroupVisitor';
import { CrudAction } from '../../enums/crudAction';

export class ScheduleFormsActionAuthVisitor extends PermissionAwareAuthGroupVisitor {
	private _serviceProvider: ServiceProvider;
	private _action: CrudAction;
	constructor(serviceProvider: ServiceProvider, action: CrudAction) {
		super();
		if (!serviceProvider) {
			throw new Error('ScheduleFormsActionAuthVisitor - ScheduleForm cannot be null.');
		}
		if (serviceProvider && !serviceProvider.service) {
			throw new Error('ScheduleFormsActionAuthVisitor - Service is not loaded in Service Provider.');
		}

		this._action = action;
		this._serviceProvider = serviceProvider;
	}

	public visitCitizen(_citizenGroup: CitizenAuthGroup): void {}

	public visitOrganisationAdmin(_userGroup: OrganisationAdminAuthGroup): void {
		switch (this._action) {
			case CrudAction.Create:
			case CrudAction.Read:
			case CrudAction.Update:
				const organisationId = this._serviceProvider.service.organisationId;
				if (_userGroup.hasOrganisationId(organisationId)) {
					this.markWithPermission();
				}
				return;
			default:
				return;
		}
	}

	public visitServiceAdmin(_userGroup: ServiceAdminAuthGroup): void {
		switch (this._action) {
			case CrudAction.Create:
			case CrudAction.Read:
			case CrudAction.Update:
				const serviceId = this._serviceProvider.serviceId;
				if (_userGroup.hasServiceId(serviceId)) {
					this.markWithPermission();
				}
				return;
			default:
				return;
		}
	}
	public visitServiceProvider(_userGroup: ServiceProviderAuthGroup): void {
		switch (this._action) {
			case CrudAction.Create:
			case CrudAction.Read:
			case CrudAction.Update:
				const serviceProviderId = this._serviceProvider.id;
				// tslint:disable-next-line: tsr-detect-possible-timing-attacks
				if (_userGroup.authorisedServiceProvider.id === serviceProviderId) {
					this.markWithPermission();
				}
				return;
			default:
				return;
		}
	}
}
export class ScheduleFormsQueryAuthVisitor extends QueryAuthGroupVisitor {
	private readonly _serviceProviderAlias: string;
	private readonly _serviceAlias: string;

	constructor(serviceAlias: string, serviceProviderAlias: string) {
		super();
		this._serviceAlias = serviceAlias;
		this._serviceProviderAlias = serviceProviderAlias;
	}

	public visitCitizen(_citizenGroup: CitizenAuthGroup): void {
		this.addAuthCondition('FALSE', {});
	}

	public visitOrganisationAdmin(_userGroup: OrganisationAdminAuthGroup): void {
		const orgIds = _userGroup.authorisedOrganisations.map((org) => org.id);
		this.addAuthCondition(`${this._serviceAlias}._organisationId IN (:...orgIds)`, { orgIds });
	}

	public visitServiceAdmin(_userGroup: ServiceAdminAuthGroup): void {
		const serviceIds = _userGroup.authorisedServices.map((s) => s.id);
		this.addAuthCondition(`${this._serviceAlias}._id IN (:...serviceIds)`, { serviceIds });
	}

	public visitServiceProvider(_userGroup: ServiceProviderAuthGroup): void {
		const serviceProviderId = _userGroup.authorisedServiceProvider.id;

		this.addAuthCondition(`${this._serviceProviderAlias}._id = :serviceProviderId`, { serviceProviderId });
	}
}
