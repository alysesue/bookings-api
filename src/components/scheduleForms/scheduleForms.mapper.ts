import { ScheduleForm, TimeOfDay, WeekDayBreak, WeekDaySchedule } from '../../models';
import { groupByKeyLastValue } from '../../tools/collections';
import { getErrorResult, getOkResult, isErrorResult, OptionalResult } from '../../errors';
import {
	ScheduleFormRequest,
	ScheduleFormResponseV1,
	ScheduleFormResponseV2,
	WeekDayBreakContract,
	WeekDayScheduleContract,
} from './scheduleForms.apicontract';
import { Inject } from 'typescript-ioc';
import { IdHasher } from '../../infrastructure/idHasher';

export class ScheduleFormsMapper {
	@Inject
	private idHasher: IdHasher;

	public mapToEntity(contract: ScheduleFormRequest, entity: ScheduleForm): OptionalResult<ScheduleForm, string[]> {
		entity.slotsDurationInMin = contract.slotsDurationInMin;
		entity.startDate = contract.startDate;
		entity.endDate = contract.endDate;
		const errors: string[] = [];
		const weekDaysContract = groupByKeyLastValue(contract.weekdaySchedules || [], (w) => w.weekday);
		entity.initWeekdaySchedules();

		for (const daySchedule of entity.weekdaySchedules) {
			if (weekDaysContract.has(daySchedule.weekDay)) {
				const dayContract = weekDaysContract.get(daySchedule.weekDay);
				dayContract.startDate = contract.startDate;
				dayContract.endDate = contract.endDate;
				const dayResult = this.setDayContractEntity(dayContract, daySchedule);
				if (isErrorResult(dayResult)) {
					errors.push(...getErrorResult(dayResult));
				}
			}
		}

		return errors.length > 0 ? { errorResult: errors } : { result: entity };
	}

	private setDayContractEntity(
		contract: WeekDayScheduleContract,
		entity: WeekDaySchedule,
	): OptionalResult<WeekDaySchedule, string[]> {
		entity.hasScheduleForm = contract.hasScheduleForm;
		entity.capacity = contract.capacity;
		entity.startDate = contract.startDate;
		entity.endDate = contract.endDate;
		const errors: string[] = [];

		try {
			entity.openTime = TimeOfDay.parse(contract.openTime);
		} catch (e) {
			errors.push(e.message);
		}

		try {
			entity.closeTime = TimeOfDay.parse(contract.closeTime);
		} catch (e) {
			errors.push(e.message);
		}

		const mappedBreaks = this.mapBreaks(contract, entity);
		if (isErrorResult(mappedBreaks)) {
			errors.push(...getErrorResult(mappedBreaks));
		} else {
			entity.breaks = getOkResult(mappedBreaks);
		}

		return errors.length > 0 ? { errorResult: errors } : { result: entity };
	}

	private mapBreaks(
		daySchedule: WeekDayScheduleContract,
		entity: WeekDaySchedule,
	): OptionalResult<WeekDayBreak[], string[]> {
		const errors: string[] = [];
		const result: WeekDayBreak[] = [];

		for (const entry of daySchedule.breaks || []) {
			let startTime: TimeOfDay = null;
			let endTime: TimeOfDay = null;

			try {
				startTime = TimeOfDay.parse(entry.startTime);
			} catch (e) {
				errors.push(e.message);
			}

			try {
				endTime = TimeOfDay.parse(entry.endTime);
			} catch (e) {
				errors.push(e.message);
			}

			if (startTime && endTime) {
				const mapped = WeekDayBreak.create(daySchedule.weekday, startTime, endTime, entity.scheduleForm);
				result.push(mapped);
			}
		}

		return errors.length > 0 ? { errorResult: errors } : { result };
	}

	public mapToResponseV1(template: ScheduleForm): ScheduleFormResponseV1 {
		if (!template) {
			return undefined;
		}

		const response = new ScheduleFormResponseV1();
		response.id = template.id;
		response.slotsDurationInMin = template.slotsDurationInMin;
		response.weekdaySchedules = template.weekdaySchedules?.map((w) => this.mapDayScheduleToResponse(w)) || [];
		response.startDate = template.startDate;
		response.endDate = template.endDate;

		return response;
	}

	public mapToResponseV2(template: ScheduleForm): ScheduleFormResponseV2 {
		if (!template) {
			return undefined;
		}

		const response = new ScheduleFormResponseV2();

		response.id = this.idHasher.encode(template.id);
		response.slotsDurationInMin = template.slotsDurationInMin;
		response.weekdaySchedules = template.weekdaySchedules?.map((w) => this.mapDayScheduleToResponse(w)) || [];
		response.startDate = template.startDate;
		response.endDate = template.endDate;

		return response;
	}

	private mapDayScheduleToResponse(daySchedule: WeekDaySchedule): WeekDayScheduleContract {
		const dayContract = new WeekDayScheduleContract();
		dayContract.weekday = daySchedule.weekDay;
		dayContract.hasScheduleForm = daySchedule.hasScheduleForm;
		dayContract.openTime = daySchedule.openTime?.toJSON();
		dayContract.closeTime = daySchedule.closeTime?.toJSON();
		dayContract.breaks = daySchedule.breaks?.map((e) => this.mapBreaksToResponse(e));
		dayContract.capacity = daySchedule.capacity;
		dayContract.startDate = daySchedule.startDate;
		dayContract.endDate = daySchedule.endDate;

		return dayContract;
	}

	private mapBreaksToResponse = (dayBreak: WeekDayBreak): WeekDayBreakContract => {
		const breakContract = new WeekDayBreakContract();
		breakContract.startTime = dayBreak.startTime.toJSON();
		breakContract.endTime = dayBreak.endTime.toJSON();

		return breakContract;
	};
}
