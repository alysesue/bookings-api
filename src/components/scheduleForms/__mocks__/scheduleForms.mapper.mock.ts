import { ScheduleForm } from '../../../models/entities';
import { ScheduleFormsMapper } from '../scheduleForms.mapper';
import { OptionalResult } from '../../../errors';
import { ScheduleFormResponseV1, ScheduleFormResponseV2 } from '../scheduleForms.apicontract';

export class ScheduleFormsMapperMock implements Partial<ScheduleFormsMapper> {
	public static mapToEntity = jest.fn<OptionalResult<ScheduleForm, string[]>, any>();
	public static mapToResponseV1 = jest.fn<ScheduleFormResponseV1, any>();
	public static mapToResponseV2 = jest.fn<ScheduleFormResponseV2, any>();

	public mapToEntity(...params): OptionalResult<ScheduleForm, string[]> {
		return ScheduleFormsMapperMock.mapToEntity(...params);
	}

	public mapToResponseV1(...params): ScheduleFormResponseV1 {
		return ScheduleFormsMapperMock.mapToResponseV1(...params);
	}

	public mapToResponseV2(...params): ScheduleFormResponseV2 {
		return ScheduleFormsMapperMock.mapToResponseV2(...params);
	}
}
