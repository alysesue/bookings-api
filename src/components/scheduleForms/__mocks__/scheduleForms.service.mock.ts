import { ScheduleFormsService } from '../scheduleForms.service';

export class ScheduleFormsServiceMock implements Partial<ScheduleFormsService> {
	public static updateScheduleFormInEntity = jest.fn();

	public async updateScheduleFormInEntity(...params): Promise<any> {
		return await ScheduleFormsServiceMock.updateScheduleFormInEntity(...params);
	}
}
