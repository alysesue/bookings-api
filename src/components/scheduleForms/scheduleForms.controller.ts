import {
	Body,
	Controller,
	Delete,
	Deprecated,
	Get,
	Path,
	Post,
	Put,
	Response,
	Route,
	SuccessResponse,
	Tags,
} from 'tsoa';
import { ScheduleFormRequest, ScheduleFormResponse } from './scheduleForms.apicontract';
import { Inject } from 'typescript-ioc';
import { ScheduleFormsService } from './scheduleForms.service';
import { MOLAuth } from 'mol-lib-common';

@Route('v1/scheduleForms')
@Tags('ScheduleForms')
export class ScheduleFormsController extends Controller {
	@Inject
	private scheduleFormService: ScheduleFormsService;

	@Post('')
	@SuccessResponse(201, 'Created')
	@MOLAuth({ admin: {} })
	@Response(401, 'Valid authentication types: [admin]')
	public async createScheduleForm(@Body() timeslot: ScheduleFormRequest): Promise<ScheduleFormResponse> {
		this.setStatus(201);
		return await this.scheduleFormService.createScheduleForm(timeslot);
	}

	@Get('')
	@MOLAuth({ admin: {} })
	@Response(401, 'Valid authentication types: [admin]')
	public async getScheduleForms(): Promise<ScheduleFormResponse[]> {
		return await this.scheduleFormService.getScheduleForms();
	}

	@Put('{id}')
	@SuccessResponse(200, 'Updated')
	@MOLAuth({ admin: {} })
	@Response(401, 'Valid authentication types: [admin]')
	public async updateScheduleForm(
		@Path() id: number,
		@Body() timeslot: ScheduleFormRequest,
	): Promise<ScheduleFormResponse> {
		return await this.scheduleFormService.updateScheduleForm(id, timeslot);
	}

	@Delete('{id}')
	@SuccessResponse(200, 'Deleted')
	@MOLAuth({ admin: {} })
	@Response(401, 'Valid authentication types: [admin]')
	public async deleteScheduleForm(@Path() id: number): Promise<any> {
		return await this.scheduleFormService.deleteScheduleForm(id);
	}
}
