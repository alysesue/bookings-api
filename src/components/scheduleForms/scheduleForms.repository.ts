import { Inject, InRequestScope } from 'typescript-ioc';
import { ScheduleForm, ServiceProvider } from '../../models';
import { DeleteResult, FindManyOptions, In, SelectQueryBuilder } from 'typeorm';
import { groupByKey } from '../../tools/collections';
import { WeekDayBreakRepository } from './weekdaybreak.repository';
import { RepositoryBase } from '../../core/repository';
import { groupByKeyLastValue } from '../../tools/collections';
import { IEntityWithScheduleForm } from '../../models/interfaces';
import { UserContext } from '../../infrastructure/auth/userContext';
import { ScheduleFormsQueryAuthVisitor } from './scheduleForms.auth';

@InRequestScope
export class ScheduleFormsRepository extends RepositoryBase<ScheduleForm> {
	@Inject
	private weekDayBreakRepo: WeekDayBreakRepository;
	@Inject
	private userContext: UserContext;

	constructor() {
		super(ScheduleForm);
	}

	private async querySelectScheduleForm(ids: number[]): Promise<SelectQueryBuilder<ScheduleForm>> {
		const idCondition = ids ? '"scheduleForm"."id" IN (:...ids)' : ' TRUE ';
		const authGroups = await this.userContext.getAuthGroups();
		const { userCondition, userParams } = await new ScheduleFormsQueryAuthVisitor(
			'SPService',
			'serviceProvider',
		).createUserVisibilityCondition(authGroups);

		const repository = await this.getRepository();
		return repository
			.createQueryBuilder('scheduleForm')
			.where(
				[userCondition, idCondition]
					.filter((c) => c)
					.map((c) => `(${c})`)
					.join(' AND '),
				{
					...userParams,
					ids,
				},
			)
			.leftJoinAndSelect('scheduleForm.weekdaySchedules', 'weekdaySchedules')
			.leftJoinAndSelect('scheduleForm.serviceProvider', 'serviceProvider')
			.leftJoinAndSelect('serviceProvider._service', 'SPService');
	}

	public async getScheduleFormById(id: number): Promise<ScheduleForm> {
		const query = await this.querySelectScheduleForm([id]);
		const scheduleForm = await query.getOne();
		return scheduleForm ? (await this.populateBreaks([scheduleForm]))[0] : null;
	}

	public async getScheduleForms(ids?: number[]): Promise<ScheduleForm[]> {
		const query = await this.querySelectScheduleForm(ids);
		const schedules = await query.getMany();
		return this.populateBreaks(schedules);
	}

	private async populateBreaks(schedules: ScheduleForm[]): Promise<ScheduleForm[]> {
		if (!schedules.length) {
			return null;
		}
		const scheduleIds = schedules.map((s) => s.id);
		const breaks = await this.weekDayBreakRepo.getBreaksForSchedules(scheduleIds);
		const breaksPerSchedule = groupByKey(breaks, (e) => e.getScheduleId());

		for (const scheduleForm of schedules) {
			const scheduleBreaks = breaksPerSchedule.get(scheduleForm.id) || [];
			scheduleForm.setBreaks(scheduleBreaks);
		}

		return schedules;
	}

	public async saveScheduleForm(scheduleForm: ScheduleForm): Promise<ScheduleForm> {
		const breaks = scheduleForm.getAllBreaks();
		const saved = await (await this.getRepository()).save(scheduleForm);

		if (scheduleForm.id > 0) {
			await this.weekDayBreakRepo.deleteBreaksForSchedule(scheduleForm.id);
		}
		const savedBreaks = await this.weekDayBreakRepo.save(breaks);

		saved.setBreaks(savedBreaks);
		return saved;
	}

	public async deleteScheduleForm(scheduleFormId: number): Promise<DeleteResult> {
		await this.weekDayBreakRepo.deleteBreaksForSchedule(scheduleFormId);
		return (await this.getRepository()).delete(scheduleFormId);
	}

	public async populateScheduleForms<T extends IEntityWithScheduleForm>(entries: T[]): Promise<T[]> {
		const scheduleFormIds = entries.map((e) => e.scheduleFormId).filter((id) => !!id);
		if (scheduleFormIds.length === 0) {
			return entries;
		}

		const scheduleFormsById = groupByKeyLastValue(await this.getScheduleForms(scheduleFormIds), (s) => s.id);

		for (const entry of entries.filter((c) => !!c.scheduleFormId)) {
			entry.scheduleForm = scheduleFormsById.get(entry.scheduleFormId);
		}
		return entries;
	}

	public async populateSingleEntryScheduleForm<T extends IEntityWithScheduleForm>(entry: T): Promise<T> {
		if (!entry) {
			return entry;
		}

		return (await this.populateScheduleForms([entry]))[0];
	}
}
