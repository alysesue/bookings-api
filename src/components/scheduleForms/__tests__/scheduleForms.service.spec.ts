import { ScheduleFormsService } from '../scheduleForms.service';
import { ScheduleFormRequest, WeekDayBreakContract, WeekDayScheduleContract } from '../scheduleForms.apicontract';
import { ScheduleFormsRepository } from '../scheduleForms.repository';
import { Container } from 'typescript-ioc';
import { Organisation, Service, ServiceProvider, User } from '../../../models';
import { Weekday } from '../../../enums/weekday';
import { MOLErrorV2 } from 'mol-lib-api-contract';
import { TimeslotsScheduleRepository } from '../../timeslotsSchedules/timeslotsSchedule.repository';
import { ScheduleFormsActionAuthVisitor } from '../scheduleForms.auth';
import { UserContext } from '../../../infrastructure/auth/userContext';
import { OrganisationAdminAuthGroup } from '../../../infrastructure/auth/authGroup';
import { AsyncFunction, TransactionManager } from '../../../core/transactionManager';
import { IsolationLevel } from 'typeorm/driver/types/IsolationLevel';
import { UserContextMock } from '../../../infrastructure/auth/__mocks__/userContext';
import { TransactionManagerMock } from '../../../core/__mocks__/transactionManager.mock';

jest.mock('../scheduleForms.auth');

afterAll(() => {
	jest.resetAllMocks();
	if (global.gc) global.gc();
});

const visitorObject = {
	hasPermission: jest.fn(),
};

const serviceMockWithTemplate = new Service();
serviceMockWithTemplate.id = 2;

const userMock = User.createAdminUser({
	molAdminId: 'd080f6ed-3b47-478a-a6c6-dfb5608a199d',
	userName: 'UserName',
	email: 'test@email.com',
	name: 'Name',
});

const organisation = new Organisation();
organisation.id = 1;

// tslint:disable-next-line: no-big-function
describe('Schedules form template services ', () => {
	beforeAll(() => {
		Container.bind(UserContext).to(UserContextMock);
		Container.bind(ScheduleFormsRepository).to(ScheduleFormsRepositoryMock);
		Container.bind(TimeslotsScheduleRepository).to(TimeslotsScheduleRepositoryMock);
		Container.bind(TransactionManager).to(TransactionManagerMock);
	});

	beforeEach(() => {
		jest.clearAllMocks();

		visitorObject.hasPermission.mockReturnValue(true);
		(ScheduleFormsActionAuthVisitor as jest.Mock).mockImplementation(() => visitorObject);
		UserContextMock.getCurrentUser.mockImplementation(() => Promise.resolve(userMock));
		UserContextMock.getAuthGroups.mockImplementation(() =>
			Promise.resolve([new OrganisationAdminAuthGroup(userMock, [organisation])]),
		);
		TransactionManagerMock.runInTransaction.mockImplementation(
			async <T extends unknown>(_isolationLevel: IsolationLevel, asyncFunction: AsyncFunction<T>): Promise<T> =>
				await asyncFunction(),
		);
	});

	it('should throw error because open and close times have wrong format', async () => {
		const entity = ServiceProvider.create('Jhon', 1);
		const saveEntity = jest.fn<Promise<typeof entity>, any>((e) => Promise.resolve(e));
		const scheduleFormsRequest: ScheduleFormRequest = {
			slotsDurationInMin: 5,
			weekdaySchedules: [
				{
					weekday: Weekday.Monday,
					hasScheduleForm: true,
					openTime: '2323',
					closeTime: '25:25',
				} as WeekDayScheduleContract,
			],
		} as ScheduleFormRequest;

		try {
			const scheduleFormsService = Container.get(ScheduleFormsService);
			await scheduleFormsService.updateScheduleFormInEntity(scheduleFormsRequest, entity, saveEntity);
		} catch (e) {
			expect(e.message).toBe('Invalid request parameters.');
			expect((e as MOLErrorV2).responseData).toMatchSnapshot();
		}
		expect(ScheduleFormsRepositoryMock.saveScheduleForm).toBeCalledTimes(0);
	});

	it('should throw error because close time have wrong format', async () => {
		const entity = ServiceProvider.create('Jhon', 1);
		const saveEntity = jest.fn<Promise<typeof entity>, any>((e) => Promise.resolve(e));
		const scheduleFormsRequest: ScheduleFormRequest = {
			slotsDurationInMin: 5,
			weekdaySchedules: [
				{
					weekday: Weekday.Monday,
					hasScheduleForm: true,
					openTime: '23:23',
					closeTime: '11:73',
				} as WeekDayScheduleContract,
			],
		} as ScheduleFormRequest;

		try {
			const scheduleFormsService = Container.get(ScheduleFormsService);
			await scheduleFormsService.updateScheduleFormInEntity(scheduleFormsRequest, entity, saveEntity);
		} catch (e) {
			expect(e.message).toBe('Invalid request parameters.');
			expect((e as MOLErrorV2).responseData).toMatchSnapshot();
		}
		expect(ScheduleFormsRepositoryMock.saveScheduleForm).toBeCalledTimes(0);
	});

	it('should throw error because openTime > closeTime', async () => {
		const entity = ServiceProvider.create('Jhon', 1);
		const saveEntity = jest.fn<Promise<typeof entity>, any>((e) => Promise.resolve(e));
		const scheduleFormsRequest: ScheduleFormRequest = {
			slotsDurationInMin: 5,
			weekdaySchedules: [
				{
					weekday: Weekday.Monday,
					hasScheduleForm: true,
					openTime: '23:23',
					closeTime: '11:23',
				} as WeekDayScheduleContract,
			],
		} as ScheduleFormRequest;

		try {
			const scheduleFormsService = Container.get(ScheduleFormsService);
			await scheduleFormsService.updateScheduleFormInEntity(scheduleFormsRequest, entity, saveEntity);
		} catch (e) {
			expect(e.message).toBe('Invalid request parameters.');
			expect((e as MOLErrorV2).responseData).toMatchSnapshot();
		}
		expect(ScheduleFormsRepositoryMock.saveScheduleForm).toBeCalledTimes(0);
	});

	it('should throw error because slotsDurationInMin < (closeTime - openTime)', async () => {
		const entity = ServiceProvider.create('Jhon', 1);
		const saveEntity = jest.fn<Promise<typeof entity>, any>((e) => Promise.resolve(e));
		const scheduleFormsRequest: ScheduleFormRequest = {
			slotsDurationInMin: 65,
			weekdaySchedules: [
				{
					weekday: Weekday.Monday,
					hasScheduleForm: true,
					openTime: '11:23',
					closeTime: '12:23',
				} as WeekDayScheduleContract,
			],
		} as ScheduleFormRequest;

		try {
			const scheduleFormsService = Container.get(ScheduleFormsService);
			await scheduleFormsService.updateScheduleFormInEntity(scheduleFormsRequest, entity, saveEntity);
		} catch (e) {
			expect(e.message).toBe('Invalid request parameters.');
			expect((e as MOLErrorV2).responseData).toMatchSnapshot();
		}
		expect(ScheduleFormsRepositoryMock.saveScheduleForm).toBeCalledTimes(0);
		expect(saveEntity).toBeCalledTimes(0);
	});

	it('should generate timeslots', async () => {
		const entity = ServiceProvider.create('Jhon', 2);
		entity.id = 1;
		entity.service = new Service();
		entity.service.id = 2;
		const saveEntity = jest.fn<Promise<typeof entity>, any>((e) => Promise.resolve(e));
		const scheduleFormsRequest: ScheduleFormRequest = {
			slotsDurationInMin: 60,
			weekdaySchedules: [
				{
					weekday: Weekday.Monday,
					hasScheduleForm: true,
					openTime: '08:30',
					closeTime: '12:30',
					breaks: [{ startTime: '11:00', endTime: '11:30' } as WeekDayBreakContract],
				} as WeekDayScheduleContract,
			],
		} as ScheduleFormRequest;

		ScheduleFormsRepositoryMock.saveScheduleForm.mockImplementation((e) => Promise.resolve(e));

		const scheduleFormsService = Container.get(ScheduleFormsService);
		await scheduleFormsService.updateScheduleFormInEntity(scheduleFormsRequest, entity, saveEntity);

		expect(ScheduleFormsRepositoryMock.saveScheduleForm).toBeCalledTimes(1);
		expect(saveEntity).toBeCalledTimes(1);

		const timeslotSchedule = entity.timeslotsSchedule;
		expect(timeslotSchedule).toBeDefined();
		expect(timeslotSchedule.timeslotItems[0]._startTime.toString()).toBe('08:30');
		expect(timeslotSchedule.timeslotItems[0]._endTime.toString()).toBe('09:30');
		expect(timeslotSchedule.timeslotItems[1]._startTime.toString()).toBe('09:30');
		expect(timeslotSchedule.timeslotItems[1]._endTime.toString()).toBe('10:30');
		expect(timeslotSchedule.timeslotItems[2]._startTime.toString()).toBe('11:30');
		expect(timeslotSchedule.timeslotItems[2]._endTime.toString()).toBe('12:30');
	});

	it('should delete old schedule form', async () => {
		const entity = ServiceProvider.create('Jhon', 2);
		entity.id = 1;
		entity.service = new Service();
		entity.service.id = 2;
		entity.scheduleFormId = 5;
		const saveEntity = jest.fn<Promise<typeof entity>, any>((e) => Promise.resolve(e));
		const scheduleFormsRequest: ScheduleFormRequest = {
			slotsDurationInMin: 60,
			weekdaySchedules: [
				{
					weekday: Weekday.Monday,
					hasScheduleForm: true,
					openTime: '08:30',
					closeTime: '12:30',
					breaks: [],
				} as WeekDayScheduleContract,
			],
		} as ScheduleFormRequest;

		ScheduleFormsRepositoryMock.saveScheduleForm.mockImplementation((e) => Promise.resolve(e));
		ScheduleFormsRepositoryMock.deleteScheduleForm.mockImplementation(() => Promise.resolve());

		const scheduleFormsService = Container.get(ScheduleFormsService);
		await scheduleFormsService.updateScheduleFormInEntity(scheduleFormsRequest, entity, saveEntity);

		expect(ScheduleFormsRepositoryMock.saveScheduleForm).toBeCalledTimes(1);
		expect(saveEntity).toBeCalledTimes(1);
		expect(ScheduleFormsRepositoryMock.deleteScheduleForm).toBeCalledWith(5);
	});

	it('should delete old timeslot schedule', async () => {
		const entity = ServiceProvider.create('Jhon', 2);
		entity.id = 1;
		entity.service = new Service();
		entity.service.id = 2;
		entity.timeslotsScheduleId = 6;
		const saveEntity = jest.fn<Promise<typeof entity>, any>((e) => Promise.resolve(e));
		const scheduleFormsRequest: ScheduleFormRequest = {
			slotsDurationInMin: 60,
			weekdaySchedules: [
				{
					weekday: Weekday.Monday,
					hasScheduleForm: true,
					openTime: '08:30',
					closeTime: '12:30',
					breaks: [],
				} as WeekDayScheduleContract,
			],
		} as ScheduleFormRequest;

		ScheduleFormsRepositoryMock.saveScheduleForm.mockImplementation((e) => Promise.resolve(e));
		TimeslotsScheduleRepositoryMock.deleteTimeslotsSchedule.mockImplementation(() => Promise.resolve());

		const scheduleFormsService = Container.get(ScheduleFormsService);
		await scheduleFormsService.updateScheduleFormInEntity(scheduleFormsRequest, entity, saveEntity);

		expect(ScheduleFormsRepositoryMock.saveScheduleForm).toBeCalledTimes(1);
		expect(saveEntity).toBeCalledTimes(1);
		expect(TimeslotsScheduleRepositoryMock.deleteTimeslotsSchedule).toBeCalledWith(6);
	});
});

class ScheduleFormsRepositoryMock implements Partial<ScheduleFormsRepository> {
	public static saveScheduleForm = jest.fn();
	public static deleteScheduleForm = jest.fn();

	public async saveScheduleForm(...params): Promise<any> {
		return await ScheduleFormsRepositoryMock.saveScheduleForm(...params);
	}

	public async deleteScheduleForm(...params): Promise<any> {
		return await ScheduleFormsRepositoryMock.deleteScheduleForm(...params);
	}
}

class TimeslotsScheduleRepositoryMock implements Partial<TimeslotsScheduleRepository> {
	public static deleteTimeslotsSchedule = jest.fn();

	public async deleteTimeslotsSchedule(...params): Promise<any> {
		return await TimeslotsScheduleRepositoryMock.deleteTimeslotsSchedule(...params);
	}
}
