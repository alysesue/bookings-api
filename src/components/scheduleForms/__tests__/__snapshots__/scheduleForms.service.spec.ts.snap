// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Schedules form template services  should throw error because close time have wrong format 1`] = `
Array [
  "Value 11:73 is not a valid time.",
]
`;

exports[`Schedules form template services  should throw error because open and close times have wrong format 1`] = `
Array [
  "Value 2323 is not a valid time.",
  "Value 25:25 is not a valid time.",
]
`;

exports[`Schedules form template services  should throw error because openTime > closeTime 1`] = `
Array [
  "Close time [11:23] must be greater than open time [23:23]",
]
`;

exports[`Schedules form template services  should throw error because slotsDurationInMin < (closeTime - openTime) 1`] = `
Array [
  "The interval between open and close times [11:23 â€” 12:23] must be greater than slot duration [65 minutes]",
]
`;
